### PR TITLE
Cover all nullable warnings

### LINKED ISSUE #29691: Move page to compiler message section
This page describes some of the warnings generated when the nullable context is set to `warning` or `enable`. To improve its use, do the following:

- [x] Make sure the list of warnings on this page is exhaustive.
- [x] Add F1 and help keywords for each `CS????` id.
- [x] Update the page to focus on how to / troubleshoot each warning.
- [x] Remove F1 and help keywords from error not documented page.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d92a07ce-fd73-47c5-513f-d1381fa9bf57
* Version Independent ID: fa3d6d5c-02a2-0654-8a46-693706376423
* Content: [Resolve nullable warnings](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-warnings)
* Content Source: [docs/csharp/nullable-warnings.md](https://github.com/dotnet/docs/blob/main/docs/csharp/nullable-warnings.md)
* Product: **dotnet-csharp**
* Technology: **csharp-null-safety**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #29691 

This PR does the following:

1. Add F1 and help viewer keywords for all C# compiler warnings enabled by `#nullable warning` (or `enable`). 
2. Add the*CSnnnn* and error message text for each warning message to increase SEO on the warnings and messages. (Note: this is why the bullet lists don't contain contractions. They match the compiler messages).
3. Move this file to the compiler messages section, and add the CSnnnn keys to the displayName property.
4. Add sections for any missing warnings that weren't covered.

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/csharp/nullable-warnings?branch=pr-en-us-30045)

### CODE REVIEW

**Comment by BillWagner** [comment]
@adegeo This is ready for review.

File: docs/csharp/language-reference/compiler-messages/nullable-warnings.md [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+You can take one of three actions to address these warnings. One is to add the `?` annotation to make the variable a nullable reference type. That change may cause other warnings. Changing a variable from a non-nullable reference to a nullable reference changes its default *null-state* from *not-null* to *maybe-null*. The compiler's static analysis may find instances where you dereference a variable that is *maybe-null*.
+
+The other actions instruct the compiler that the right-hand-side of the assignment is *not-null*. The expression on the right-hand-side could be null-checked before assignment, as shown in the following example:
+
+:::code language="csharp" source="./snippets/null-warnings/Program.cs" id="NullGuard":::
+
+The previous examples demonstrate assignment to the return value of a method. You may annotate the method (or property) to indicate when a method returns a not-null value. The <xref:System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute?displayProperty=nameWithType> often specifies that a return value is *not-null* when an input argument is *not-null*. Another alternative is to add the null forgiving operator, `!` to the right-hand side:
```
Comment (adegeo) [inline docs/csharp/language-reference/compiler-messages/nullable-warnings.md:191]:
```suggestion
The previous examples demonstrate assignment of the return value of a method. You may annotate the method (or property) to indicate when a method returns a not-null value. The <xref:System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute?displayProperty=nameWithType> often specifies that a return value is *not-null* when an input argument is *not-null*. Another alternative is to add the null forgiving operator, `!` to the right-hand side:
```

File: docs/csharp/language-reference/compiler-messages/nullable-warnings.md [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+The compiler produces a warning because the `message` parameter is assigned `null` *and* the method returns `true`. The `NotNullWhen` attribute indicates that shouldn't happen.
+
+To address these warnings, update your code so it matches the expectations of the attributes you've applied. You may change the attributes, or the algorithm.
+
+## Exhaustive switch expression
+
+Switch expressions must be *exhaustive*, meaning that all input values must be handled. Even for non-nullable reference types, the `null` value must be accounted for. The compiler issues warnings when the null value isn't handled:
+
+- **CS8655** # WRN_SwitchExpressionNotExhaustiveForNull: The switch expression does not handle some null inputs (it is not exhaustive). For example, the pattern '{0}' is not covered.
+- **CS8847** # WRN_SwitchExpressionNotExhaustiveForNullWithWhen: The switch expression does not handle some null inputs (it is not exhaustive). For example, the pattern '{0}' is not covered. However, a pattern with a 'when' clause might successfully match this value.
```
Comment (adegeo) [inline docs/csharp/language-reference/compiler-messages/nullable-warnings.md:316]: I think there was a copy paste error here. Updating to match the style of the others
```suggestion
- **CS8655** - *The switch expression does not handle some null inputs (it is not exhaustive). For example, the pattern '{0}' is not covered.*
- **CS8847** - *The switch expression does not handle some null inputs (it is not exhaustive). For example, the pattern '{0}' is not covered. However, a pattern with a 'when' clause might successfully match this value.*
```

**Review by adegeo — APPROVED** [review]
Approved. Some minor things to consider/fix. Great job on this content!

File: docs/csharp/language-reference/compiler-messages/nullable-warnings.md [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+The compiler produces a warning because the `message` parameter is assigned `null` *and* the method returns `true`. The `NotNullWhen` attribute indicates that shouldn't happen.
+
+To address these warnings, update your code so it matches the expectations of the attributes you've applied. You may change the attributes, or the algorithm.
+
+## Exhaustive switch expression
+
+Switch expressions must be *exhaustive*, meaning that all input values must be handled. Even for non-nullable reference types, the `null` value must be accounted for. The compiler issues warnings when the null value isn't handled:
+
+- **CS8655** - *The switch expression does not handle some null inputs (it is not exhaustive). For example, the pattern '{0}' is not covered.*
+- **CS8847** - *The switch expression does not handle some null inputs (it is not exhaustive). For example, the pattern '{0}' is not covered. However, a pattern with a 'when' clause might successfully match this value.*
```
Comment (BillWagner) [inline docs/csharp/language-reference/compiler-messages/nullable-warnings.md:316]: Remove the holes:

```suggestion
- **CS8655** - *The switch expression does not handle some null inputs (it is not exhaustive).*
- **CS8847** - *The switch expression does not handle some null inputs (it is not exhaustive). However, a pattern with a 'when' clause might successfully match this value.*
```

**Pull request merged**